diff --git a/src/apple-llvm/src/projects/libtapi/CMakeLists.txt b/src/apple-llvm/src/projects/libtapi/CMakeLists.txt
index 9e7a3d8..5458fc1 100644
--- a/src/apple-llvm/src/projects/libtapi/CMakeLists.txt
+++ b/src/apple-llvm/src/projects/libtapi/CMakeLists.txt
@@ -32,9 +32,9 @@ configure_file(
 
 macro(add_tapi_library name)
   cmake_parse_arguments(ARG
-    "SHARED"
+    "SHARED;CACHE"
     ""
-    "ADDITIONAL_HEADERS"
+    "INTERNAL;LINK_LIBS;ADDITIONAL_HEADERS"
     ${ARGN})
   set(srcs)
   if(MSVC_IDE OR XCODE)
@@ -70,10 +70,12 @@ macro(add_tapi_library name)
     )
   endif()
 
-  if(ARG_SHARED)
+  #if(ARG_SHARED)
     set(ARG_ENABLE_SHARED SHARED)
-  endif()
-  llvm_add_library(${name} ${ARG_ENABLE_SHARED} ${ARG_UNPARSED_ARGUMENTS} ${srcs})
+  #endif()
+
+  add_library(${name} ${ARG_ENABLE_SHARED} ${ARG_UNPARSED_ARGUMENTS} ${srcs})
+  set_target_properties(${name} PROPERTIES LINKER_LANGUAGE CXX)
 
   if(TARGET ${name})
     target_link_libraries(${name} ${cmake_2_8_12_INTERFACE} ${LLVM_COMMON_LIBS})
