diff --git a/bsd-user/freebsd/target_os_clock.h b/bsd-user/freebsd/target_os_clock.h
new file mode 100644
index 0000000000..4af8143beb
--- /dev/null
+++ b/bsd-user/freebsd/target_os_clock.h
@@ -0,0 +1,24 @@
+#pragma once
+
+#define TARGET_CLOCK_REALTIME		0
+#define TARGET_CLOCK_VIRTUAL		1
+#define TARGET_CLOCK_PROF		2
+#define TARGET_CLOCK_MONOTONIC		4
+#define TARGET_CLOCK_UPTIME_FAST	8
+#define TARGET_CLOCK_UPTIME		5
+#define TARGET_CLOCK_UPTIME_PRECISE	7
+#define TARGET_CLOCK_REALTIME_PRECISE	9
+#define TARGET_CLOCK_REALTIME_FAST	10
+#define TARGET_CLOCK_MONOTONIC_PRECISE	11
+#define TARGET_CLOCK_MONOTONIC_FAST	12
+#define TARGET_CLOCK_SECOND		13
+#define TARGET_CLOCK_THREAD_CPUTIME_ID	14
+#define	TARGET_CLOCK_PROCESS_CPUTIME_ID 15
+#define	TARGET_CLOCK_BOOTTIME		CLOCK_UPTIME
+#define	TARGET_CLOCK_REALTIME_COARSE	CLOCK_REALTIME_FAST
+#define	TARGET_CLOCK_MONOTONIC_COARSE	CLOCK_MONOTONIC_FAST
+
+#define TARGET_TIMER_RELTIME	0x0	/* relative timer */
+#define TARGET_TIMER_ABSTIME	0x1	/* absolute timer */
+
+#define	TARGET_UMTX_ABSTIME		0x01
diff --git a/bsd-user/freebsd/target_os_sysctl.h b/bsd-user/freebsd/target_os_sysctl.h
new file mode 100644
index 0000000000..d74cec87a3
--- /dev/null
+++ b/bsd-user/freebsd/target_os_sysctl.h
@@ -0,0 +1,52 @@
+#pragma once
+
+/* sysctl constants */
+#if defined(__linux__)
+#define        CTL_MAXNAME     24      /* largest number of components supported */
+
+#define        CTLTYPE         0xf     /* mask for the type */
+#define        CTLTYPE_NODE    1       /* name is a node */
+#define        CTLTYPE_INT     2       /* name describes an integer */
+#define        CTLTYPE_STRING  3       /* name describes a string */
+#define        CTLTYPE_S64     4       /* name describes a signed 64-bit number */
+#define        CTLTYPE_OPAQUE  5       /* name describes a structure */
+#define        CTLTYPE_STRUCT  CTLTYPE_OPAQUE  /* name describes a structure */
+#define        CTLTYPE_UINT    6       /* name describes an unsigned integer */
+#define        CTLTYPE_LONG    7       /* name describes a long */
+#define        CTLTYPE_ULONG   8       /* name describes an unsigned long */
+#define        CTLTYPE_U64     9       /* name describes an unsigned 64-bit number */
+#define        CTLTYPE_U8      0xa     /* name describes an unsigned 8-bit number */
+#define        CTLTYPE_U16     0xb     /* name describes an unsigned 16-bit number */
+#define        CTLTYPE_S8      0xc     /* name describes a signed 8-bit number */
+#define        CTLTYPE_S16     0xd     /* name describes a signed 16-bit number */
+#define        CTLTYPE_S32     0xe     /* name describes a signed 32-bit number */
+#define        CTLTYPE_U32     0xf     /* name describes an unsigned 32-bit number */
+
+#define        CTLFLAG_RD      0x80000000      /* Allow reads of variable */
+#define        CTLFLAG_WR      0x40000000      /* Allow writes to the variable */
+#define        CTLFLAG_RW      (CTLFLAG_RD|CTLFLAG_WR)
+#define        CTLFLAG_DORMANT 0x20000000      /* This sysctl is not active yet */
+#define        CTLFLAG_ANYBODY 0x10000000      /* All users can set this var */
+#define        CTLFLAG_SECURE  0x08000000      /* Permit set only if securelevel<=0 */
+#define        CTLFLAG_PRISON  0x04000000      /* Prisoned roots can fiddle */
+#define        CTLFLAG_DYN     0x02000000      /* Dynamic oid - can be freed */
+#define        CTLFLAG_SKIP    0x01000000      /* Skip this sysctl when listing */
+#define        CTLMASK_SECURE  0x00F00000      /* Secure level */
+#define        CTLFLAG_TUN     0x00080000      /* Default value is loaded from getenv() */
+#define        CTLFLAG_RDTUN   (CTLFLAG_RD|CTLFLAG_TUN)
+#define        CTLFLAG_RWTUN   (CTLFLAG_RW|CTLFLAG_TUN)
+#define        CTLFLAG_MPSAFE  0x00040000      /* Handler is MP safe */
+#define        CTLFLAG_VNET    0x00020000      /* Prisons with vnet can fiddle */
+#define        CTLFLAG_DYING   0x00010000      /* Oid is being removed */
+#define        CTLFLAG_CAPRD   0x00008000      /* Can be read in capability mode */
+#define        CTLFLAG_CAPWR   0x00004000      /* Can be written in capability mode */
+#define        CTLFLAG_STATS   0x00002000      /* Statistics, not a tuneable */
+#define        CTLFLAG_NOFETCH 0x00001000      /* Don't fetch tunable from getenv() */
+#define        CTLFLAG_CAPRW   (CTLFLAG_CAPRD|CTLFLAG_CAPWR)
+
+#define CTL_SYSCTL_NAME 1
+
+/* ... and others */
+#define VM_OVERCOMMIT 12
+
+#endif
diff --git a/bsd-user/linux/os-sys.c b/bsd-user/linux/os-sys.c
index 12af7d94ca..3e4c4f97dd 100644
--- a/bsd-user/linux/os-sys.c
+++ b/bsd-user/linux/os-sys.c
@@ -45,6 +45,7 @@
 #include "target_arch_sysarch.h"
 #include "target_os_vmparam.h"
 #include "target_os_user.h"
+#include "target_os_sysctl.h"
 
 #include "os-dev_proc.h"
 
@@ -107,16 +108,275 @@ patch_osrelease(char *osp, off_t at, size_t olen) {
     memcpy(osp + QEMU_REL_POS, QEMU_REL_SUF, sizeof(QEMU_REL_SUF));
 }
 
-void
-init_bsd_sysctl(void)
+// Handlers for each case
+static int
+handle_hw_pagesizes(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
 {
-    patch_osrelease(osrelease_str, 0, sizeof(osrelease_str));
-    patch_osrelease(kversion_str, sizeof(FREEBSD_ABI_VENDOR), sizeof(kversion_str));
-    for (abi_ulong *ps = pagesizes; (void *)ps < ((void *)pagesizes+sizeof(pagesizes)); ps++) {
-        *ps = tswapal(*ps);
+    *holdlen = sizeof(pagesizes);
+    if (oldlen) {
+        if (*holdlen > oldlen)
+            *holdlen = oldlen;
+        memcpy(holdp, pagesizes, *holdlen);
+    }
+    return 0; // success
+}
+
+static int
+handle_hw_availpages(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    *holdlen = sizeof(abi_long);
+    if (oldlen) {
+        struct meminfo_entry se = get_meminfo_value("MemAvailable");
+        if (se.error) {
+            return -TARGET_EINVAL;
+        }
+        *((abi_long *)holdp) = se.value / PAGE_SIZE;
+    }
+    return 0; // success
+}
+
+static int
+handle_kern_ps_strings(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return -TARGET_EINVAL;
+}
+
+static int
+handle_hw_machine_arch(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    struct utsname buffer;
+    const char *cp;
+
+    if (uname(&buffer) == -1) {
+        return -TARGET_EINVAL;
+    }
+    if (strcmp(buffer.machine, "x86_64") == 0) {
+        cp = "amd64";
+    } else {
+        cp = buffer.machine;
+    }
+    *holdlen = strlen(cp) + 1;
+    if (oldlen) {
+        if (*holdlen > oldlen)
+            *holdlen = oldlen;
+        memcpy(holdp, cp, *holdlen);
+    }
+    return 0;
+}
+
+static inline int sysctlnametomib(const char *name, int *mibp, size_t *sizep);
+static int
+handle_sysctl_name2oid(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    size_t sizep = hnewplen / sizeof(int);
+    int r = sysctlnametomib(hnewp, holdp, &sizep);
+    *holdlen = sizep * sizeof(int);
+    return r;
+}
+
+static int oidfmt(const int *oid, int len, char *fmt, int fmtlen, uint32_t *kind);
+static int
+handle_sysctl_oidfmt(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return oidfmt(
+        hnewp,
+        hnewplen / sizeof(int),
+        holdp == 0 ? 0 : holdp + sizeof(int),
+        hnewplen < 4 ? 0 : hnewplen - sizeof(int),
+        holdp
+    );
+}
+
+static int
+handle_sysctl_name(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return oidfmt(
+        hnewp,
+        hnewplen / sizeof(int),
+        holdp,
+        oldlen,
+        0
+    );
+}
+
+static int
+handle_generic_const_int(void *holdp, size_t oldlen, size_t *holdlen, int val)
+{
+    if (oldlen < sizeof(int)) {
+        *holdlen = sizeof(int);
+        return -TARGET_ENOMEM;
+    }
+    *(int*)holdp = val;
+    *holdlen = sizeof(int);
+    return 0;
+}
+
+static int
+handle_vm_overcommit(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return handle_generic_const_int(holdp, oldlen, holdlen, 0);
+}
+
+static int
+handle_kern_sched_cpusetsizemin(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return handle_generic_const_int(holdp, oldlen, holdlen, 1);
+}
+
+static int
+handle_kern_sched_cpusetsize(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return handle_generic_const_int(holdp, oldlen, holdlen, 1);
+}
+
+static int
+handle_dummy(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t hnewplen)
+{
+    return -TARGET_ENOENT;
+}
+
+struct handler {
+    const char *name;
+    int (*func)(void *holdp, size_t oldlen, size_t *holdlen, const void *hnewp, size_t nnewlen);
+    unsigned int kind;
+    bool prefix;
+    size_t nmib;
+    int mib[8];
+};
+
+/*
+ * Procedure for adding a new handler:
+ * - fill out the table here
+ * - implement the handler
+ * - add the name parser to sysctlnametomib
+ */
+static struct handler handlers[] = {
+    {"hw.pagesizes", handle_hw_pagesizes, CTLFLAG_RW},
+    {"hw.availpages", handle_hw_availpages, CTLFLAG_RD},
+    {"kern.ps_strings", handle_kern_ps_strings, CTLFLAG_RD | CTLFLAG_MPSAFE},
+    {"kern.ostype", handle_dummy, CTLFLAG_RD | CTLFLAG_CAPRD},
+    {"kern.hostname", handle_dummy, CTLTYPE_STRING | CTLFLAG_RW | CTLFLAG_PRISON | CTLFLAG_CAPRD | CTLFLAG_MPSAFE},
+    {"kern.osrelease", handle_dummy, CTLTYPE_STRING | CTLFLAG_RW | CTLFLAG_CAPRD | CTLFLAG_MPSAFE},
+    {"kern.osreldate", handle_dummy, CTLTYPE_INT | CTLFLAG_CAPRD | CTLFLAG_RD | CTLFLAG_MPSAFE},
+    {"kern.arandom", handle_dummy, CTLTYPE_OPAQUE | CTLFLAG_CAPRD | CTLFLAG_RD | CTLFLAG_MPSAFE},
+    {"kern.version", handle_dummy, CTLTYPE_STRING | CTLFLAG_MPSAFE | CTLFLAG_RD},
+    {"kern.usrstack", handle_dummy, CTLTYPE_ULONG | CTLFLAG_MPSAFE | CTLFLAG_RD | CTLFLAG_CAPRD},
+    {"kern.sched.cpusetsize", handle_kern_sched_cpusetsize, CTLTYPE_UINT | CTLFLAG_RD | CTLFLAG_CAPRD},
+    {"kern.sched.cpusetsizemin", handle_kern_sched_cpusetsizemin, CTLTYPE_UINT | CTLFLAG_RD | CTLFLAG_CAPRD},
+    {"hw.machine", handle_hw_machine_arch, CTLTYPE_STRING | CTLFLAG_RD | CTLFLAG_CAPRD | CTLFLAG_MPSAFE},
+    {"hw.machine_arch", handle_hw_machine_arch, CTLTYPE_STRING | CTLFLAG_RD | CTLFLAG_CAPRD | CTLFLAG_MPSAFE},
+    {"vm.overcommit", handle_vm_overcommit, CTLTYPE_INT | CTLFLAG_RW},
+    {"sysctl.name2oid", handle_sysctl_name2oid, CTLTYPE_INT | CTLFLAG_RW | CTLFLAG_ANYBODY | CTLFLAG_MPSAFE | CTLFLAG_CAPRW},
+    {"sysctl.name", handle_sysctl_name, CTLFLAG_RD | CTLFLAG_MPSAFE | CTLFLAG_CAPRD, true},
+    {"sysctl.oidfmt", handle_sysctl_oidfmt, CTLFLAG_RD | CTLFLAG_MPSAFE | CTLFLAG_CAPRD, true},
+    {NULL, NULL} // Terminator
+};
+
+#if defined(__linux__)
+
+enum extra_oids {
+    KERN_SCHED = 0x70000000,
+    KERN_SCHED_CPUSETSIZEMIN,
+    KERN_SCHED_CPUSETSIZE,
+    CTL_VM,
+    HW_PAGESIZES,
+    HW_AVAILPAGES,
+};
+
+static inline int sysctlnametomib(const char *name, int *mibp, size_t *sizep)
+{
+#define NAME(c, v, children) \
+    else if (strncmp(name, (c), sizeof(c)-1) == 0 && (name[sizeof(c)-1] == 0 || name[sizeof(c)-1] == '.')) { \
+        if (seen < *sizep) { \
+            mibp[seen] = (v); \
+        } \
+        name += sizeof(c); \
+        seen += 1; \
+        if (name[-1] == 0) { \
+            *sizep = seen; \
+            return 0; \
+        } \
+        children \
+    }
+
+    int seen = 0;
+    const char *orig_name = name;
+
+    if (0) {}
+    NAME("sysctl", CTL_SYSCTL,
+        NAME("name2oid", CTL_SYSCTL_NAME2OID, )
+        NAME("oidfmt", CTL_SYSCTL_OIDFMT, )
+        NAME("name", CTL_SYSCTL_NAME, )
+    )
+    NAME("kern", CTL_KERN,
+        NAME("ps_strings", KERN_PS_STRINGS, )
+        NAME("ostype", KERN_OSTYPE, )
+        NAME("hostname", KERN_HOSTNAME, )
+        NAME("osrelease", KERN_OSRELEASE, )
+        NAME("osreldate", KERN_OSRELDATE, )
+        NAME("arandom", KERN_ARND, )
+        NAME("version", KERN_VERSION, )
+        NAME("usrstack", KERN_VERSION, )
+        NAME("sched", KERN_SCHED,
+            NAME("cpusetsizemin", KERN_SCHED_CPUSETSIZEMIN, )
+            NAME("cpusetsize", KERN_SCHED_CPUSETSIZE, )
+        )
+    )
+    NAME("hw", CTL_HW,
+        NAME("pagesizes", HW_PAGESIZES, )
+        NAME("availpages", HW_AVAILPAGES, )
+        NAME("machine", HW_MACHINE, )
+        NAME("machine_arch", HW_MACHINE_ARCH, )
+    )
+    NAME("vm", CTL_VM,
+        NAME("overcommit", VM_OVERCOMMIT, )
+    )
+	qemu_log("sysctlnametomib(%s): not implemented yet\n", orig_name);
+	return -TARGET_ENOENT;
+}
+
+static inline int sysctl(const int *name, u_int namelen, void *oldp, size_t *oldlenp,
+                         const void *newp, size_t newlen)
+{
+    for (int i = 0; handlers[i].name; i++) {
+        if ((handlers[i].prefix ? handlers[i].nmib <= namelen : handlers[i].nmib == namelen) && memcmp(handlers[i].mib, name, handlers[i].nmib * sizeof(int)) == 0) {
+            if (handlers[i].prefix) {
+                newp = &name[handlers[i].nmib];
+                newlen = (namelen - handlers[i].nmib) * sizeof(int);
+            }
+            return handlers[i].func(oldp, *oldlenp, oldlenp, newp, newlen);
+        }
     }
+#if DEBUG
+    printf("Unimplemented sysctl ");
+    for (int i = 0; i < namelen; i++) {
+        if (i != 0) {
+            printf(",");
+        }
+        printf("%#010x", name[i]);
+    }
+    printf("\n");
+#endif
+    return -TARGET_ENOENT;
 }
 
+static inline int sysctlbyname(const char *name, void *oldp, size_t *oldlenp,
+         const void *newp, size_t newlen)
+{
+    for (int i = 0; handlers[i].name; i++) {
+        if (strcmp(name, handlers[i].name) == 0) {
+            return handlers[i].func(oldp, *oldlenp, oldlenp, newp, newlen);
+        }
+    }
+#if DEBUG
+    printf("Unimplemented sysctlbyname %s\n", name);
+#endif
+    return -TARGET_ENOENT;
+}
+
+#endif
+
+
 #ifdef TARGET_ABI32
 /*
  * Limit the amount of available memory to be most of the 32-bit address
@@ -1105,7 +1365,7 @@ host_to_target_vfc_flags(int flags)
  * sysctl, see /sys/kern/kern_sysctl.c:sysctl_sysctl_oidfmt() (compare to
  * src/sbin/sysctl/sysctl.c)
  */
-static int oidfmt(int *oid, int len, char *fmt, uint32_t *kind)
+static int oidfmt(const int *oid, int len, char *fmt, int fmtlen, uint32_t *kind)
 {
 #if !defined(__linux__)
     int qoid[CTL_MAXNAME + 2];
@@ -1128,10 +1388,31 @@ static int oidfmt(int *oid, int len, char *fmt, uint32_t *kind)
     }
 
     if (fmt) {
-        strcpy(fmt, (char *)(buf + sizeof(uint32_t)));
+        strncpy(fmt, (char *)(buf + sizeof(uint32_t)), fmtlen);
     }
     return 0;
 #else
+    for (int i = 0; handlers[i].name != 0; i++) {
+        if ((handlers[i].prefix ? len >= handlers[i].nmib : len == handlers[i].nmib) && memcmp(oid, handlers[i].mib, handlers[i].nmib * sizeof(int)) == 0) {
+            if (kind) {
+                *kind = handlers[i].kind;
+            }
+            if (fmt) {
+                strncpy(fmt, handlers[i].name, fmtlen);
+            }
+            return 0;
+        }
+    }
+#if DEBUG
+    printf("Unimplemented sysctl (oidfmt) ");
+    for (int i = 0; i < len; i++) {
+        if (i != 0) {
+            printf(",");
+        }
+        printf("%#010x", oid[i]);
+    }
+    printf("\n");
+#endif
     return -TARGET_ENOENT;
 #endif
 }
@@ -1245,26 +1526,6 @@ static inline void sysctl_oidfmt(uint32_t *holdp)
     holdp[0] = tswap32(holdp[0]);
 }
 
-#if defined(__linux__)
-static inline int sysctlnametomib(const char *name, int *mibp, size_t *sizep)
-{
-	qemu_log("sysctlnametomib(%s): not implemented yet\n", name);
-	return -TARGET_ENOENT;
-}
-
-static inline int sysctlbyname(const char *name, void *oldp, size_t *oldlenp,
-         const void *newp, size_t newlen)
-{
-	abort();
-}
-
-static inline int sysctl(const int *name, u_int namelen, void *oldp, size_t *oldlenp,
-	const void *newp, size_t newlen)
-{
-	return -TARGET_ENOENT;
-}
-#endif
-
 static abi_long do_freebsd_sysctl_oid(CPUArchState *env, int32_t *snamep,
         int32_t namelen, void *holdp, size_t *holdlenp, void *hnewp,
         size_t newlen)
@@ -1280,7 +1541,7 @@ static abi_long do_freebsd_sysctl_oid(CPUArchState *env, int32_t *snamep,
     const char *ret_str;
 
     holdlen = oldlen = *holdlenp;
-    oidfmt(snamep, namelen, NULL, &kind);
+    oidfmt(snamep, namelen, NULL, 0, &kind);
 
     /* Handle some arch/emulator dependent sysctl()'s here. */
     switch (snamep[0]) {
@@ -1477,8 +1738,6 @@ static abi_long do_freebsd_sysctl_oid(CPUArchState *env, int32_t *snamep,
             ret = 0;
             goto out;
         }
-#else
-	abort();
 #endif
     }
     break;
@@ -1725,7 +1984,7 @@ static abi_long do_freebsd_sysctl_oid(CPUArchState *env, int32_t *snamep,
     }
 #ifdef DEBUG
     else {
-        printf("sysctl(mib[0]=%d, mib[1]=%d, mib[3]=%d...) returned %d\n",
+        printf("sysctl(mib[0]=%d, mib[1]=%d, mib[2]=%d...) returned %d\n",
                snamep[0], snamep[1], snamep[2], (int)ret);
     }
 #endif
@@ -1746,76 +2005,6 @@ out_str:
 }
 
 
-struct handler {
-    const char *name;
-    int (*func)(void *holdp, size_t oldlen, size_t *holdlen);
-};
-
-// Handlers for each case
-static int
-handle_hw_pagesizes(void *holdp, size_t oldlen, size_t *holdlen)
-{
-    *holdlen = sizeof(pagesizes);
-    if (oldlen) {
-        if (*holdlen > oldlen)
-            *holdlen = oldlen;
-        memcpy(holdp, pagesizes, *holdlen);
-    }
-    return 0; // success
-}
-
-static int
-handle_hw_availpages(void *holdp, size_t oldlen, size_t *holdlen)
-{
-    *holdlen = sizeof(abi_long);
-    if (oldlen) {
-        struct meminfo_entry se = get_meminfo_value("MemAvailable");
-        if (se.error) {
-            return -TARGET_EINVAL;
-        }
-        *((abi_long *)holdp) = se.value / PAGE_SIZE;
-    }
-    return 0; // success
-}
-
-static int
-handle_kern_ps_strings(void *holdp, size_t oldlen, size_t *holdlen)
-{
-    return -TARGET_EINVAL;
-}
-
-static int
-handle_hw_machine_arch(void *holdp, size_t oldlen, size_t *holdlen)
-{
-    struct utsname buffer;
-    const char *cp;
-
-    if (uname(&buffer) == -1) {
-        return -TARGET_EINVAL;
-    }
-    if (strcmp(buffer.machine, "x86_64") == 0) {
-        cp = "amd64";
-    } else {
-        cp = buffer.machine;
-    }
-    *holdlen = strlen(cp) + 1;
-    if (oldlen) {
-        if (*holdlen > oldlen)
-            *holdlen = oldlen;
-        memcpy(holdp, cp, *holdlen);
-    }
-    return 0;
-}
-
-static const struct handler handlers[] = {
-    {"hw.pagesizes", handle_hw_pagesizes},
-    {"hw.availpages", handle_hw_availpages},
-    {"kern.ps_strings", handle_kern_ps_strings},
-    {"hw.machine", handle_hw_machine_arch},
-    {"hw.machine_arch", handle_hw_machine_arch},
-    {NULL, NULL} // Terminator
-};
-
 /*
  * This syscall was created to make sysctlbyname(3) more efficient, but we can't
  * really provide it in bsd-user.  Notably, we must always translate the names
@@ -1861,7 +2050,7 @@ abi_long do_freebsd_sysctlbyname(CPUArchState *env, abi_ulong namep,
 
     for (const struct handler *h = handlers; h->name; h++) {
         if (namelen == strlen(h->name) && strcmp(snamep, h->name) == 0) {
-            ret = h->func(holdp, oldlen, &holdlen);
+            ret = h->func(holdp, oldlen, &holdlen, hnewp, newlen);
 	    goto out;
         }
     }
@@ -1898,6 +2087,10 @@ abi_long do_freebsd_sysctl(CPUArchState *env, abi_ulong namep, int32_t namelen,
     size_t holdlen;
     abi_ulong oldlen = 0;
     int32_t *snamep = g_malloc(sizeof(int32_t) * namelen), *p, *q, i;
+    if (namelen == 0) {
+        ret = -TARGET_EINVAL;
+        goto out;
+    }
 
     /* oldlenp is read/write, pre-check here for write */
     if (oldlenp) {
@@ -1950,3 +2143,19 @@ abi_long do_freebsd_sysarch(void *cpu_env, abi_long arg1, abi_long arg2)
 {
     return do_freebsd_arch_sysarch(cpu_env, arg1, arg2);
 }
+
+void
+init_bsd_sysctl(void)
+{
+    patch_osrelease(osrelease_str, 0, sizeof(osrelease_str));
+    patch_osrelease(kversion_str, sizeof(FREEBSD_ABI_VENDOR), sizeof(kversion_str));
+    for (abi_ulong *ps = pagesizes; (void *)ps < ((void *)pagesizes+sizeof(pagesizes)); ps++) {
+        *ps = tswapal(*ps);
+    }
+    for (int i = 0; handlers[i].name; i++) {
+        handlers[i].nmib = 8;
+        if (sysctlnametomib(handlers[i].name, &handlers[i].mib[0], &handlers[i].nmib) < 0) {
+            printf("Handler we can't convert to mib: %s\n", handlers[i].name);
+        }
+    }
+}
diff --git a/bsd-user/linux/os-thread.c b/bsd-user/linux/os-thread.c
index 45639057fa..f81124b68a 100644
--- a/bsd-user/linux/os-thread.c
+++ b/bsd-user/linux/os-thread.c
@@ -24,6 +24,7 @@
 #include "signal-common.h"
 #include "target_arch_cpu.h"
 #include "target_arch_thread.h"
+#include "target_os_clock.h"
 #include "tcg/startup.h"
 #include "exec/tb-flush.h"
 #include "qemu/guest-random.h"
@@ -175,6 +176,63 @@ void *new_freebsd_thread_start(void *arg)
 /*
  * FreeBSD user mutex (_umtx) emulation
  */
+static void make_ts_relative(struct timespec *tm, clockid_t clockid) {
+    struct timespec now;
+    assert(clock_gettime(clockid, &now) == 0);
+
+    /* if the target is in the past, return zero */
+    if (tm->tv_sec < now.tv_sec || (tm->tv_sec == now.tv_sec && tm->tv_nsec < now.tv_nsec)) {
+        tm->tv_sec = 0;
+        tm->tv_nsec = 0;
+        return;
+    }
+
+    /* borrowing */
+    if (tm->tv_nsec < now.tv_nsec) {
+        tm->tv_sec -= 1;
+        tm->tv_nsec += 1000000000;
+    }
+    /* simple subtraction */
+    tm->tv_sec -= now.tv_sec;
+    tm->tv_nsec -= now.tv_nsec;
+}
+
+static void translate_timeout(void *t, size_t tsz, struct timespec *timeout, struct timespec **timeoutp, int *op_flags) {
+    if (tsz == 0) {
+        *timeoutp = NULL;
+        return;
+    }
+
+    if (tsz == sizeof(struct timespec)) {
+        /* just a timespec */
+        memcpy(timeout, t, tsz);
+        *timeoutp = timeout;
+        return;
+    }
+
+    /* _umtx_time */
+    assert(tsz >= sizeof(struct timespec) + sizeof(uint32_t) + sizeof(uint32_t));
+    memcpy(timeout, t, sizeof(struct timespec));
+    *timeoutp = timeout;
+    uint32_t flags = *(uint32_t*)(t + sizeof(struct timespec));
+    uint32_t clock = *(uint32_t*)(t + sizeof(struct timespec) + sizeof(uint32_t));
+    clockid_t host_clockid;
+
+    switch (clock) {
+        case TARGET_CLOCK_REALTIME:
+        case TARGET_CLOCK_REALTIME_FAST:
+        case TARGET_CLOCK_REALTIME_PRECISE:
+        *op_flags |= FUTEX_CLOCK_REALTIME;
+        host_clockid = CLOCK_REALTIME;
+        break;
+        default:
+        host_clockid = CLOCK_MONOTONIC;
+    }
+
+    if (flags & TARGET_UMTX_ABSTIME) {
+        make_ts_relative(timeout, host_clockid);
+    }
+}
 static int tcmpset_al(volatile abi_ulong *addr, abi_ulong a, abi_ulong b)
 {
 #if !defined(__linux__)
@@ -259,9 +317,11 @@ static abi_long _umtx_wait_uint(uint32_t *addr, uint32_t target_val,
         } while (1);
     } else
 #endif
-    assert(tsz == 0 && t == 0);
-    return get_errno(safe_futex(addr, QEMU_UMTX_OP(UMTX_OP_WAIT_UINT),
-            target_val, 0, 0, 0));
+    struct timespec *timeoutp;
+    struct timespec timeout;
+    int op = QEMU_UMTX_OP(UMTX_OP_WAIT_UINT);
+    translate_timeout(t, tsz, &timeout, &timeoutp, &op);
+    return get_errno(safe_futex(addr, op, target_val, timeoutp, 0, 0));
 }
 
 abi_long freebsd_umtx_wait_uint(abi_ulong obj, uint32_t target_val,
@@ -307,9 +367,11 @@ static abi_long _umtx_wait_uint_private(uint32_t *addr, uint32_t target_val,
         } while (1);
     }
 #endif /* DETECT_DEADLOCK */
-    assert (t == 0 && tsz == 0); /* "Not implemented yet: open PR!" */
-    return get_errno(safe_futex(addr, QEMU_UMTX_OP(UMTX_OP_WAIT_UINT_PRIVATE),
-			    target_val, NULL, NULL, 0));
+    struct timespec *timeoutp;
+    struct timespec timeout;
+    int op = QEMU_UMTX_OP(UMTX_OP_WAIT_UINT_PRIVATE);
+    translate_timeout(t, tsz, &timeout, &timeoutp, &op);
+    return get_errno(safe_futex(addr, op, target_val, NULL, NULL, 0));
 }
 
 abi_long freebsd_umtx_wait_uint_private(abi_ulong obj, uint32_t target_val,
@@ -355,8 +417,11 @@ static abi_long _umtx_wait(uint32_t *addr, uint32_t target_val, size_t tsz,
         } while (1);
     } else
 #endif /* DETECT_DEADLOCK */
-    assert (t == 0 && tsz == 0); /* "Not implemented yet: open PR!" */
-    return get_errno(safe_futex(addr, QEMU_UMTX_OP(UMTX_OP_WAIT), target_val, NULL, NULL, 0));
+    struct timespec *timeoutp;
+    struct timespec timeout;
+    int op = QEMU_UMTX_OP(UMTX_OP_WAIT);
+    translate_timeout(t, tsz, &timeout, &timeoutp, &op);
+    return get_errno(safe_futex(addr, op, target_val, NULL, NULL, 0));
 }
 
 abi_long freebsd_umtx_wait(abi_ulong targ_addr, abi_ulong target_id, size_t tsz,
diff --git a/bsd-user/linux/target_os_clock.h b/bsd-user/linux/target_os_clock.h
new file mode 120000
index 0000000000..76199036d9
--- /dev/null
+++ b/bsd-user/linux/target_os_clock.h
@@ -0,0 +1 @@
+../freebsd/target_os_clock.h
\ No newline at end of file
diff --git a/bsd-user/linux/target_os_sysctl.h b/bsd-user/linux/target_os_sysctl.h
new file mode 120000
index 0000000000..0417f4566b
--- /dev/null
+++ b/bsd-user/linux/target_os_sysctl.h
@@ -0,0 +1 @@
+../freebsd/target_os_sysctl.h
\ No newline at end of file
diff --git a/bsd-user/linux/truss.c b/bsd-user/linux/truss.c
index 4feafd0cbc..5603021218 100644
--- a/bsd-user/linux/truss.c
+++ b/bsd-user/linux/truss.c
@@ -12,6 +12,7 @@
 
 static void alloc_syscall(TaskState *t, int num)
 {
+    /*
     assert(t->in_syscall == 0);
     assert(t->cs.number == 0);
     assert(t->cs.sc == NULL);
@@ -19,6 +20,7 @@ static void alloc_syscall(TaskState *t, int num)
     for (u_int i = 0; i < nitems(t->cs.s_args); i++) {
         assert(t->cs.s_args[i] == NULL);
     }
+    */
     memset(t->cs.args, 0, sizeof(t->cs.args));
     t->cs.number = num;
     t->cs.sc = &decoded_syscalls[num];
@@ -33,7 +35,9 @@ static void free_syscall(TaskState *t)
         free(t->cs.s_args[i]);
     }
     memset(&t->cs, 0, sizeof(t->cs));
-    fclose(t->outfile);
+    if (t->outfile) {
+        fclose(t->outfile);
+    }
     t->outfile = NULL;
     t->in_syscall = 0;
 }
@@ -117,6 +121,10 @@ static void print_syscall_ret(TaskState *ts, abi_ulong ret, abi_ulong ret2, int
     char **s_args;
     int i, len, nargs;
 
+    if (!ts->cs.sc) {
+        printf("???\n");
+        return;
+    }
     name = ts->cs.sc->name;
     nargs = ts->cs.nargs;
     s_args = ts->cs.s_args;
diff --git a/bsd-user/signal.c b/bsd-user/signal.c
index b3d8381d29..27e3616f96 100644
--- a/bsd-user/signal.c
+++ b/bsd-user/signal.c
@@ -117,7 +117,6 @@ int host_to_target_signal(int sig)
         case HOST_SIGCONT: return TARGET_SIGCONT;
         case HOST_SIGCHLD: return TARGET_SIGCHLD;
         case HOST_SIGIO:   return TARGET_SIGIO;
-        case HOST_SIGPWR: return TARGET_SIGTHR;
     }
     if (sig >= HOST_SIGRTMIN && sig <= HOST_SIGRTMAX) {
         sig += (HOST_SIGRTMIN - TARGET_SIGRTMIN);
