diff --git a/src/fuser.c b/src/fuser.c
index d3622d6..6e4fc83 100644
--- a/src/fuser.c
+++ b/src/fuser.c
@@ -38,7 +38,10 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <sys/socket.h>
+#if defined(__linux__)
 #include <sys/sysmacros.h>
+#include <mntent.h>
+#endif
 #include <sys/wait.h>
 #include <arpa/inet.h>
 #include <netinet/in.h>
@@ -47,7 +50,6 @@
 #include <dirent.h>
 #include <unistd.h>
 #include <ctype.h>
-#include <mntent.h>
 #include <signal.h>
 #include <getopt.h>
 #include <setjmp.h>
